### DISCUSSION TITLE
Stale snapshot images

### DISCUSSION THREAD

**Discussion by 00madz** [djdd87/SynoAI#138] [Q&A]
Hi - firstly just wanted to say thanks so much for putting this great project together - I went through the process of setting it up last night and the documentation is some of the best i've come across! 
I think I have all the necessary parts configured and my camera is responding to motion and triggering synoai - the issue I am having is that the snapshots it is being fed from SSS seem to be stale. I don't think this is a iframe issue - it's a HikVision camera and I'm just using standard H.264 with 10fps and an i-frame interval of 10. The snapshots though are completely stale - eg when I walked in front of the camera this morning to trigger it, the image retrieved is still from last night. 
I increased the MaxSnapshots to 3 and all three images returned are identical.
I changed the Quality from Balanced to Low so the images would be coming from the camera substream, not sure if that matters? That is what I record constantly in SS though - I only switch to High Quality for 1 minute when an event occurs.
Have you seen anything like this before?
Is there a way I can call this API manually to see what the synology is serving up?

Here is my config file:

```
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Warning"
    }
  },

  "Url": "http://synology:5000",
  "User": "xxxxxx",
  "Password": "xxxxxx",
  "Quality": "Low",
  "MinSizeX": 50,
  "MinSizeY": 50,
  "Delay": 5000,
  "DelayAfterSuccess": 5000,
  "MaxSnapshots": 3,
  "DrawMode": "Matches",
  "SaveOriginalSnapshot": "Always",
  "DaysToKeepCaptures": 14,
  "AI": {
    "Type": "DeepStack",
    "Url": "http://synology:7250"
  },

  "Notifiers": [
    {
      "Type": "Webhook",
      "Url": "http://homeassistant:8123/api/webhook/synoai_motion_feed",
      "SendImage": "True"
    }
  ],

  "Cameras": [
    {
      "Name": "Driveway",
      "Types": [ "Person", "Car" ],
      "Threshold": 45
    },
    {
      "Name": "Front Courtyard",
      "Types": [ "Person" ],
      "Threshold": 45
    }
  ]
}
```

And log snippet (I don't get timestamps in the log for some reason??):
```
info: SynoAI.Controllers.CameraController[0]
      Captures Clean Up: Cleaning up images older than 14 day(s).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 1 of 3 requested at EVENT TIME 0ms.
info: SynoAI.Services.SynologyService[0]
      Front Courtyard: Taking snapshot
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot received in 4167ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 1 of 3 received at EVENT TIME 4167ms.
info: SynoAI.Services.AIService[0]
      Front Courtyard: DeepStackAI: Processed successfully (4188ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 1 of 3 contains 0 objects at EVENT TIME 8356ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Saving original image
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Saving image to 'Captures/Front Courtyard/Front Courtyard_2022_07_28_11_17_23_655_Original.jpeg'.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Imaged saved to 'Captures/Front Courtyard/Front Courtyard_2022_07_28_11_17_23_655_Original.jpeg' (6ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Nothing detected by the AI at EVENT TIME 8366ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Finished (8366ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 2 of 3 requested at EVENT TIME 8366ms.
info: SynoAI.Services.SynologyService[0]
      Front Courtyard: Taking snapshot
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot received in 4149ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 2 of 3 received at EVENT TIME 12515ms.
info: SynoAI.Services.AIService[0]
      Front Courtyard: DeepStackAI: Processed successfully (170ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 2 of 3 contains 0 objects at EVENT TIME 12686ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Saving original image
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Saving image to 'Captures/Front Courtyard/Front Courtyard_2022_07_28_11_17_27_986_Original.jpeg'.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Imaged saved to 'Captures/Front Courtyard/Front Courtyard_2022_07_28_11_17_27_986_Original.jpeg' (6ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Nothing detected by the AI at EVENT TIME 12696ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Finished (12696ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 3 of 3 requested at EVENT TIME 12696ms.
info: SynoAI.Services.SynologyService[0]
      Front Courtyard: Taking snapshot
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot received in 4155ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 3 of 3 received at EVENT TIME 16851ms.
info: SynoAI.Services.AIService[0]
      Front Courtyard: DeepStackAI: Processed successfully (185ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 3 of 3 contains 0 objects at EVENT TIME 17036ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Saving original image
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Saving image to 'Captures/Front Courtyard/Front Courtyard_2022_07_28_11_17_32_337_Original.jpeg'.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Imaged saved to 'Captures/Front Courtyard/Front Courtyard_2022_07_28_11_17_32_337_Original.jpeg' (8ms).
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Nothing detected by the AI at EVENT TIME 17049ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Finished (17049ms).
```

**Comment by 00madz** [comment]
Okay this is very strange, I've had a play with the syno API and using this call:

`http://server:5000/webapi/entry.cgi?version=9&name=%22Side%20Road%22&api=%22SYNO.SurveillanceStation.Camera%22&method=%22GetSnapshot%22&profileType=0&sid=xxxx`

I get fresh snapshot images for profileType 0 (High Quality) or 1 (Balanced) but as soon as I specify profileType=2, regardless of which camera I pull from, and i've tried several, 2 different brands - I get a stale image from last night??? 

Very confused

**Reply by djdd87** [reply]
I'm afraid I can't offer much help and it's not an issue I can reproduce. This proves that it's not SynoAI's fault. It's either an issue with SSS or your camera(s). You'd be best off asking on the synology forums/reddit.

**Reply by 00madz** [reply]
Yeah just thought you might have come across it before. I have contacted Synology, they have provided this initial response, but since it only seems to be me affected, I'm not holding my breath:

`Thank you for contacting Synology support. Regarding the WebAPI issue that you mentioned, it looks like it's a bug or an error in the document. We will double-check with the developer and keep you posted.`

Will post the eventual resolution here in case anyone else runs into same issue

**Reply by djdd87** [reply]
That would be great thanks. In the mean time, I'd just use high quality images if I were you anyway. You'll get better detection rates.

**Comment by 00madz** [comment]
@djdd87 the problem I have with using the higher-res snapshot images, which do work via the API, is that I seem to miss the action by about 5-10 seconds, so I just get an empty image sent. 

For example, if I walk into the frame of the camera, it triggers the motion event (i can tell as the camera flashes) I stand there for a few seconds, then walk back off. SynoAI reports no motion and when I review the captured image, it's an empty scene. Looking at the camera time watermarked into the image vs the recorded footage, I can see that the snapshot is about 10 seconds too late.... 

Troubleshooting this is part of the reason I was hoping to see timestamps in the log file (?) as I don't understand what is causing the delay. I see in the message text there is mention of EVENT time - from my log above, does it appear it's taking 4 seconds to actually get the snapshot image? Is that what that means? Any ideas why it would take so long, is there a setting I can change?? when I try the request in the browser, it's instant, even for the high-res file

Thanks

**Reply by djdd87** [reply]
There are time stamps in the log file, so I'm unsure why you're not seeing them. I'd send you a screenshot of mine, but Synology's Docker app is just crap and refuses to show me a log half the time and today is one of those times.

Looking at that log above, it looks like it's running 3 detections at once - have you set up 3 action rules or something? Having 3 simultaneous calls each doing 3 calls is bound to slow things down a bit.

From a SynoAI point of view, there's very little delay. The process is simply:

- Fetch snapshot when told to do so
- Pass snapshot to AI
- Check tolerances
- Send notifications.

**Reply by djdd87** [reply]
Also when you say when you make the request in the browser are you actually getting an image back?

**Reply by djdd87** [reply]
Finally got the logs to load showing the date/times:

![image](https://user-images.githubusercontent.com/22151993/181488790-f15ee135-40c1-46f3-99e9-05188d397b45.png)

**Reply by 00madz** [reply]
I'm using portainer to view the logs, so maybe it is cutting the timestamp off, although it is not there even when I select download logs. 
![image](https://user-images.githubusercontent.com/20661478/181488603-d930ab06-07db-4dc7-a919-4ef6418c1b34.png)

Re the 3 calls - I changed MaxSnapshots to 3 trying to troubleshoot the stale image... I can set that back to 1 for the high-res, but it appears the 3 are being retrieved in order rather than at once and each snapshot has a slightly later timestamp.

When it says this in the logs:

```
info: SynoAI.Services.SynologyService[0]
      Front Courtyard: Taking snapshot
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot received in 4149ms.
info: SynoAI.Controllers.CameraController[0]
      Front Courtyard: Snapshot 2 of 3 received at EVENT TIME 12515ms.
```
Does that imply it took 4 seconds to receive the snapshot back from the API call?
And the EVENT TIME there, is that elapsed time since the webhook notification to SynoAI from Synology ??

Thanks !

**Reply by 00madz** [reply]
Hm strange about the timestamp!! I'll try through synology docker and see....
Re the browser, yeah I'm just doing a call to get the SID and then a GET for the snapshot and get a pic back:
![image](https://user-images.githubusercontent.com/20661478/181489646-c4b21fe6-ffbb-44bc-848e-6033ac05cfad.png)

**Reply by djdd87** [reply]
There should be an option in portainer to show timestamps:

https://docs.portainer.io/user/docker/containers/logs

**Reply by 00madz** [reply]
Oh wow, i've been looking right over that option! Thanks!!

**Reply by djdd87** [reply]
MaxSnapshots is a sequential loop. So it pulls the snapshot once, then if it finds no match it checks again and so on and so forth. I misread your logs, I thought it was doing 3 calls all at the same time and repeating it 3 times at once.

**Reply by djdd87** [reply]
It certainly seems like it's taking ~4 seconds to get the snapshot, yeah:

> info: SynoAI.Controllers.CameraController[0]
>       Captures Clean Up: Cleaning up images older than 14 day(s).
> info: SynoAI.Controllers.CameraController[0]
>       Front Courtyard: Snapshot 1 of 3 requested at EVENT TIME 0ms.
> info: SynoAI.Services.SynologyService[0]
>       Front Courtyard: Taking snapshot
> info: SynoAI.Controllers.CameraController[0]
>       Front Courtyard: Snapshot received in 4167ms.
> info: SynoAI.Controllers.CameraController[0]
>       Front Courtyard: Snapshot 1 of 3 received at EVENT TIME **4167ms**.

**Reply by 00madz** [reply]
So here I am walking out - note time in bottom right of image
![image](https://user-images.githubusercontent.com/20661478/181493363-96a4f6c0-0225-4cb6-872d-0d95b8640a2a.png)
Walked to middle and back now at 21:16:55
![image](https://user-images.githubusercontent.com/20661478/181493511-80158c6d-0d12-43ff-974a-9bff953a9069.png)

This is the snapshot loaded by SynoAI - taken from time: 21:17:02
![image](https://user-images.githubusercontent.com/20661478/181493673-875323cd-04b9-4790-83fc-d83e4846cdd2.png)

I'm gone by then. 

The corresponding logs:
![image](https://user-images.githubusercontent.com/20661478/181493937-6a8355fa-8e7d-402f-a7a8-09611773d635.png)

So this sort of delay isn't normal obviously then?
And it's not influenced by any of the delay settings in the syno config? 

It wouldn't have anything to do with this pre-recording buffer would it? Do you have that set on yours?

![image](https://user-images.githubusercontent.com/20661478/181494352-ba76166f-cb08-4abb-9cc0-bd3224f15c14.png)


Thanks for your help - much appreciated. I'd love to get this working.

**Reply by djdd87** [reply]
The only thing that should delay it is this as far as I know, because then SSS will wait until there's at least X seconds of motion before calling the action rules. There will, of course, be a delay on the SSS when it decides to call the action rules that's outside of my control.

![image](https://user-images.githubusercontent.com/22151993/181494970-384e4eed-76a9-4974-ab43-48945de8ab10.png)

Also your logs don't marry up with the timestamps of those images. I'm guessing you're 10 hrs ahead of UTC wherever you are. So according to those logs SynoAI was called at 21:17:19, but your photos are all well before then, so it doesn't make sense.

I'm processing 2k images and it's taking no longer than about 2 seconds most of the time. 

What is your CPU priority set to on the container? Could you try upping that?

**Reply by 00madz** [reply]
I'll give that a try re increasing the CPU priority - it's just normal/default

Yeah the watermark is based on the camera's time and i think because I have them firewalled off the internet they cant contact time server - i thought i had them pointed at synology for that, but looks like this camera is out

I did notice in your logs that some of the time to retrieve snapshots was up around 2.5 to 3s ? Seems a long time when the get request in the browser is quite quick, although I guess it sets up session and also writes out the file within that time?

**Reply by djdd87** [reply]
Can you also try just using the IP of `server` instead of the hostname in the config in case the DNS resolution on the docker side is slowing things down?

**Reply by djdd87** [reply]
You could also try opening a bash terminal and running `apt-get update` to ensure everything's up to date. I seem to have gotten some quicker timings down near half a second since running that, so potentially I may need to update the base image:

![image](https://user-images.githubusercontent.com/22151993/181508965-6739d49c-4def-4e04-8f90-ce6fd7929928.png)

**Reply by djdd87** [reply]
![image](https://user-images.githubusercontent.com/22151993/181509506-36bc53d5-4d11-4597-b464-0683b98450c4.png)

**Reply by 00madz** [reply]
So the apt-get update didnt really do much - nothing needed updating.
You were write about the dns resolution though! That was the issue - i think it's because I run a pihole as primary dns but it's not accessible from the docker host itself so I guess that means the primary dns always fails over to secondary - anyway with the IP address in place it's now a few hundred ms to get the snapshots and is working great - thanks so much for working through that with me - have shouted you a well earned coffee!! I do have one further question for you, but I'll ask in a separate post as it's unrelated.

**Reply by djdd87** [reply]
Thank goodness for that! I'll update the README troubleshooting section with this information. Thank you very much for the coffee 👍.
